atomisticparsers/h5md/parser.py

```diff
@@ -0,0 +1,787 @@
+                        else:
+                            unit = None
+                            if hasattr(val, 'units'):
+                                unit = val.units
+                                val = val.magnitude
+                            data_h5md['x_h5md_custom_calculations'].append(CalcEntry(kind=key, value=val, unit=unit))
+
+            for observable_type, observable_dict in calculation_info.items():
+                for key, observable in observable_dict.items():
+                    map_key = observable_type + '-' + key if key else observable_type
+                    obs_index = system_map[frame].get(map_key)
+                    if obs_index:
+                        val = observable.get('value', [None] * (obs_index + 1))[obs_index]
+                        if 'energ' in observable_type:  # TODO check for energies or energy when matching name
+                            if key in Energy.__dict__.keys():
+                                data['energy'][key] = dict(value=val)
+                            else:
+                                data_h5md['x_h5md_energy_contributions'].append(EnergyEntry(kind=map_key, value=val))
+                        else:
+                            if key == '':
+                                key = observable_type
+                            else:
+                                key = map_key
+
+                            if key in BaseCalculation.__dict__.keys():
+                                data[key] = val
+                            else:
+                                unit = None
+                                if hasattr(val, 'units'):
+                                    unit = val.units
+                                    val = val.magnitude
+                                data_h5md['x_h5md_custom_calculations'].append(CalcEntry(kind=map_key, value=val, unit=unit))
+
+            self.parse_thermodynamics_step(data)
+            sec_calc = sec_run.calculation[-1]
+            if format_times(sec_calc.time) != frame:  # TODO check this comparison
+                sec_calc = sec_run.m_create(Calculation)
+                sec_calc.time = frame * time_unit
+            for calc_entry in data_h5md['x_h5md_custom_calculations']:
+                sec_calc.x_h5md_custom_calculations.append(calc_entry)
+            sec_energy = sec_calc.energy
+            if not sec_energy:
+                sec_energy = sec_calc.m_create(Energy)
+            for energy_entry in data_h5md['x_h5md_energy_contributions']:
+                sec_energy.x_h5md_energy_contributions.append(energy_entry)
+
+    def parse_system(self):
+        sec_run = self.archive.run[-1]
+
+        system_info = self.system_info.get('system')
+        if not system_info:
+            self.logger.error('No particle information found in H5MD file.')
+
+        n_frames = len(system_info.get('times', []))
+        self._system_time_map = {}
+        self._system_step_map = {}
```

Comment:
i think you do not need this mapping, just provide
```python
self.trajectory_steps = list(range(n_frames))
self.thermodynamic_steps = [frame for frame in range(frame) if system_info.get('times, [None] * (frame + 1)}) is not None]

Reply:
Each observable in `parse_calculation()` can have it's own individual list of trajectory times or steps. I think I need the mapping to make the complete, non-overlapping list of steps and then assign the system_refs properly, no? I do not see how I could do that simply with the lists that you defined.

Reply:
Thermo quantities are at different steps? Then this will not work unless you set self.thermodynamic_steps each each time you parse a quantity.

Reply:
No but I think this will work, thermodynamic_steps will simply be a list of all the steps for which a thermo property is calculated then set the property to None if not calculated at that step. Not sure if I read the situation clearl.y

Reply:
ok sorry, i think i understand now. the difficulty is that observable_info and system_info are saved as dict with key name, but if instead have time_step as key, we do not need to have an extra transformation and all the mapping. I do not like the the two nested loops in parse_calculation.

Reply:
I suggest, we simply have observable_dict expressed in terms of step rather than having to do these transformations?

Reply:
ok, I guess I do not know what you mean then. Here I am expressing the dictionary now by time_step. `observable_dict` is coming essentially directly from the hdf5 file. There the configurational observables **have** to be stored in an array. So, at some point I have to make this transformation to storing by time_step. I can also do this within the observable_info() function, but it will effectively be the same in terms of the nested loops. Am I missing something here?

Reply:
actually, I am working on redoing the parsing of system_info and observable_info, so I will get back to you on this in the next round

Reply:
> There the configurational observables **have** to be stored in an array

Do you really have to store this in an array or you can already implement the mapping here?

Reply:
I completely redid this part and also the parsing of the system_info. I think it is an improvement and at least closer to what you wanted. Let me know what you think though.